### DISCUSSION TITLE
Best practice for categorizing features / fixes from pull requests on generated release notes

### DISCUSSION THREAD

**Discussion by jmontTP** [rfennell/AzurePipelines#1580] [Q&A]
Hello,

I'm currently trying to categorize the release notes based on which PRs address fixes vs. which PRs address features, specific for an Azure devops yaml pipeline. Taking the starting template for generating the release notes as a starting point, I've came up with this for categorizing pull requests for Bug items addressing fixes:

{{#forEach workItems"}}
{{#if (eq (lookup this.fields 'System.WorkItemType') 'Bug')}}
{{#forEach this.relations}}
{{#if (contains this.attributes.name 'Pull Request')}}
{{#with (lookup_a_pullrequest ../../pullRequests this.url)}}
 {{this.title}}
{{/with}}
{{/if}}
{{/forEach}}
{{/if}}
{{/forEach}}

Does an okay job as a starting point, but I would also like to sort the titles of the pull requests being iterated based on the title, as well as break out when a PR in relation to the work item gets found to prevent duplicate listings. I have the extension for each_with_sort_by_field.js, but I would need to save all pull requests found, and then sort them out when the main foreach completes before listing them out. Would it be possible to get guided on how to best perform that action? Would making a custom handlebar extension facilitate this? I'm new to this and still picking up the ropes. Any help would be greatly appreciated!

**Comment by rfennell** [comment]
Can you not use the Pull Request array as an easier starting point?

```
## Associated Pull Requests ({{pullRequests.length}})
{{#forEach pullRequests}}
* **[{{this.pullRequestId}}]({{replace (replace this.url "_apis/git/repositories" "_git") "pullRequests" "pullRequest"}})** {{this.title}}
* Associated Work Items
{{#forEach this.associatedWorkitems}}
   {{#with (lookup_a_work_item ../../relatedWorkItems this.url)}}
    - [{{this.id}}]({{replace this.url "_apis/wit/workItems" "_workitems/edit"}}) - {{lookup this.fields 'System.Title'}}
   {{/with}}
{{/forEach}}
* Associated Commits (this includes commits on the PR source branch not associated directly with the build)
{{#forEach this.associatedCommits}}
    - [{{this.commitId}}]({{this.remoteUrl}}) -  {{this.comment}}
{{/forEach}}
{{/forEach}}
```

But if you need to loop in a different way, you can make your custom extension as complex as you like, it is just javascript and can manipulate any content you provide it,

**Reply by jmontTP** [reply]
My plan is to take all the pull requests and categorize them based on the work items associated with it. If it has a 'Bug' work item, then it would be categorized as a "Fix", and so on. On the generated notes, I would then want to display them in a list based on the categories, sorted by their title names. I can see the pull request array as a better starting point, but I would have to cache those up into their own respective arrays before beginning to display them in the release notes. Is there a way to make a local array for caching those PRs based on their associated work item through Handlebars? I could then use those arrays to display their titles sorted by your custom extension.

**Reply by jmontTP** [reply]
Sidenote, I wasn't able to get a list of the work items associated with the pull requests based on the previous code you shared. Removing the #with displays the .id only, everything else is blank. I'm back on trying to make this work through the work items and fetching their PRs, no success as of yet...

**Reply by rfennell** [reply]
The PR logic in this task is something I created, it is not based on a standard Microsoft API call (as non exists). It tries to work out PR associations based on commit merge types. So there is the chance is not presenting the data needed for your use case.

I would have at https://github.com/rfennell/AzurePipelines/blob/663e7b2ea0f32a5dbec8be120696767e917a9bff/Extensions/XplatGenerateReleaseNotes/XplatGenerateReleaseNotesTask/src/ReleaseNotesFunctions.ts#L229 and see if it is even getting the information you require

**Reply by jmontTP** [reply]
As you stated, it doesn't present the data needed. I'm starting to leverage just as is, generating all pull request titles, categorized by the work item types like the original code is structured, and then manually going over the release notes to remove duplicates and sort them out before releasing. Still researching Handlebars to familiarize myself more in it and see if I can figure it out somehow instead.

**Reply by jmontTP** [reply]
Ultimately, got a powershell script sorting out and cleaning up duplicates from the generated file. Much easier to implement
